contrib/epee/include/net/net_helper.h

```diff
@@ -110,25 +111,34 @@ namespace net_utils
 			catch(...) { /* ignore */ }
 		}
 
+		inline void set_ssl(epee::net_utils::ssl_support_t ssl_support = epee::net_utils::e_ssl_support_autodetect, const std::string &private_key = std::string(), const std::string &certificate = std::string(), const std::list<std::string> &allowed_certificates = std::list<std::string>(), bool allow_any_cert = false)
+		{
+			if (ssl_support == epee::net_utils::e_ssl_support_disabled)
+				m_ctx = {boost::asio::ssl::context(boost::asio::ssl::context::sslv23), {}};
+			else
+				m_ctx = create_ssl_context(private_key, certificate, allowed_certificates, allow_any_cert);
+		}
+
     inline
-      bool connect(const std::string& addr, int port, std::chrono::milliseconds timeout, bool ssl = false, const std::string& bind_ip = "0.0.0.0")
+      bool connect(const std::string& addr, int port, std::chrono::milliseconds timeout, const std::string& bind_ip = "0.0.0.0")
     {
-      return connect(addr, std::to_string(port), timeout, ssl, bind_ip);
+      return connect(addr, std::to_string(port), timeout, bind_ip);
     }
 
     inline
-			bool connect(const std::string& addr, const std::string& port, std::chrono::milliseconds timeout, bool ssl = false, const std::string& bind_ip = "0.0.0.0")
+			bool connect(const std::string& addr, const std::string& port, std::chrono::milliseconds timeout, const std::string& bind_ip = "0.0.0.0")
 		{
-      m_connected = false;
-      m_ssl = ssl;
+			m_connected = false;
 			try
 			{
 				m_ssl_socket.next_layer().close();
 
 				// Set SSL options
 				// disable sslv2
-				m_ctx.set_options(boost::asio::ssl::context::default_workarounds | boost::asio::ssl::context::no_sslv2);
-				m_ctx.set_default_verify_paths();
+				m_ctx.context.set_options(boost::asio::ssl::context::default_workarounds | boost::asio::ssl::context::no_sslv2);
+				m_ctx.context.set_default_verify_paths();
+				m_ssl_socket.~stream();
```

Comment:
Assign a default constructed object to `m_ssl_socket`. The current code might be implementation defined.

Reply:
Ah. I found why:

/home/user/src/bitmonero/contrib/epee/include/net/net_helper.h:140:18: error: object of type 'boost::asio::ssl::stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> > >' cannot be assigned because its copy assignment operator is implicitly deleted

Reply:
If either the destructor or constructor throws, `m_ssl_socket` is in an indeterminate state and UB will occur the next time the object is touched (during destruction, for instance). It's fairly tricky to code around this - this is yet-another sneakily weird edge case within the language.

I know something similar is done in `epee`, but its something that should likely be updated too.

Reply:
It seems like there has to be a better way to reset state of the object, if not `boost::optional` will do what you want. Or you'll manually have to use `std::aligned_storage` and use some bool to track state.

Reply:
That all sounds very unlikely, no ?

Reply:
The documentation does not list any possible exceptions, however neither function is marked as `noexcept`. I'm not sure about the ASIO developer, but several Boost developers view `noexcept` as a permanent API specification. Therefore the lack of `noexcept` means some future release could throw. And I'm not entirely certain that these _can't_ throw now; the constructor and destructor sequences are platform specific (due to internal `boost::asio::ip::tcp::socket`).

 If there is no other good technique for resetting state (there doesn't appear to be), UB can be removed with heap allocation, `boost::optional`, or a custom `std::aligned_storage`. Leaving the potential of UB doesn't seem worth it considering the fixes are relatively easy, and two of them can avoid additional allocations.

Reply:
OK, I'll add a boost::optional here then.

Reply:
Actually, a shared_ptr seems even better AFAICT. I'll do that and see if anything complains.

Reply:
Ah, nevermind, that's what you said via heap allocation above. So I'll do that.